hvsock_test.go

```diff
@@ -0,0 +1,627 @@
+			"the sockets can deal with really long inputs. Look at this key mashing: " +
+			"sdflhsdfgkjdhskljjsad;kljfasd;lfkjsadl ;fasdjfopiwej09q34iur092\"i4o[piwajfliasdkf-012ior]-" +
+			"01oi3;'lSD<Fplkasdjgoisaefjoiasdlj\"hgfoaisdkf';laksdjdf[poaiseefk-0923i4roi3qwjrf9" +
+			"08sEJKEFOLIsaejf[09saEJFLKSADjf;lkasdjf;kljaslddhgaskghk"),
+		{0x5c, 0xbd, 0xb5, 0xe7, 0x6b, 0xcb, 0xe7, 0x23, 0xff, 0x7a, 0x19, 0x77, 0x2c, 0xca, 0xab, 0x3b},
+	} {
+		f.Add(b)
+	}
+
+	f.Fuzz(func(t *testing.T, a []byte) {
+		if string(a) == "" {
+			t.Skip("skipping empty string")
+		}
+		t.Logf("testing %q (%d)", a, len(a))
+		u := newUtil(t)
+		cl, sv, _ := clientServer(u)
+
+		svCh := make(chan struct{})
+		go func() {
+			defer close(svCh)
+
+			n, err := cl.Write(a)
+			u.Must(err, "client write")
+			u.Assert(n == len(a), "client did not send full message")
+			t.Log("client sent")
+
+			b := make([]byte, len(a)+5) // a little extra to make sure nothing else is sent
+			n, err = cl.Read(b)
+			u.Must(err, "cl read")
+			u.Assert(n == len(a), "client did not read full message")
+			bn := b[:n]
+			u.Assert(string(a) == string(bn), fmt.Sprintf("payload mismatch %q != %q", a, bn))
+			t.Log("client received")
+		}()
+
+		clCh := make(chan struct{})
+		go func() {
+			defer close(clCh)
+
+			b := make([]byte, len(a)+5) // a little extra to make sure nothing else is sent
+			n, err := sv.Read(b)
+			u.Must(err, "server read")
+			u.Assert(n == len(a), "server did not read full message")
+			bn := b[:n]
+			u.Assert(string(a) == string(bn), fmt.Sprintf("payload mismatch %q != %q", a, bn))
+			t.Log("server received")
+
+			n, err = sv.Write(bn)
+			u.Must(err, "server write")
+			u.Assert(n == len(bn), "server did not send full message")
+			t.Log("server sent")
+		}()
+		u.Wait(svCh, 250*time.Millisecond)
+		u.Wait(clCh, 250*time.Millisecond)
+	})
+}
+
+//
+// helpers
+//
+
+type testUtil struct {
+	T testing.TB
+}
+
+func newUtil(t testing.TB) testUtil {
+	return testUtil{
+		T: t,
+	}
+}
+
+// checks stops execution if testing failed in another go-routine
+func (u testUtil) Check() {
+	if u.T.Failed() {
+		u.T.FailNow()
+	}
+}
+
+func (u testUtil) Assert(b bool, msgs ...string) {
+	if b {
+		return
+	}
+	u.T.Helper()
+	u.T.Fatalf(_msgJoin(msgs, "failed assertion"))
+}
+
+func (u testUtil) Is(err, target error, msgs ...string) {
+	if errors.Is(err, target) {
+		return
+	}
+	u.T.Helper()
+	u.T.Fatalf(_msgJoin(msgs, "got error %q; wanted %q"), err, target)
+}
+
+func (u testUtil) Must(err error, msgs ...string) {
+	if err == nil {
+		return
+	}
+	u.T.Helper()
+	u.T.Fatalf(_msgJoin(msgs, "%v"), err)
+}
+
+func (u testUtil) Wait(ch <-chan struct{}, d time.Duration, msgs ...string) {
+	t := time.NewTimer(d)
+	defer t.Stop()
+	select {
+	case <-ch:
+	case <-t.C:
+		u.T.Helper()
+		u.T.Fatalf(_msgJoin(msgs, "timed out after %v"), d)
+	}
+}
+
+func _msgJoin(pre []string, s string) string {
```

Comment:
Any reason for the underscore? Generally not very idiomatic

Reply:
partially old habit, but also, since `hvsock_test.go` shares the same (testing) namespace for all of `winio`, I wanted to keep this hidden to prevent conflicts
I haven't come up with a good way to do so ....
(In my defense, they use underscores in the [go source code](https://cs.opensource.google/search?q=_%5Ba-z%5D&ss=go%2Fgo), and for auto-generated mkwinsyscall code....)

Reply:
TIL, looks like they use it in tests a bit and for syscall definitions they don't wanna export but want to somewhat keep the screaming case for, _BLAH_BLAH.

Reply:
I removed the leading `_`, hopefully we can isolate the hvsock tests (or testing support) in the future